src/ch/ch_monitor.c

```diff
@@ -563,25 +564,294 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+            ret = -1;
+        }
+
+        *end = tmp;
+        buf = end;
+    }
+
+    /*
+     * If the buffer still has incomplete data, lets
+     * push it to the beginning.
+     */
+    sz = buf - mon->buffer;
+    if (sz < mon->buf_fill_sz) {
+        mon->buf_offset = mon->buf_fill_sz - sz;
+        memcpy(mon->buffer, buf, sz);
+    } else {
+        mon->buf_offset = 0;
+    }
+
+    return ret;
+}
+
+static int virCHMonitorReadProcessEvents(virCHMonitorPtr mon)
+{
+    size_t max_sz = CH_MONITOR_BUFFER_SZ - mon->buf_offset;
+    char *buf = mon->buffer + mon->buf_offset;
+    virDomainObjPtr vm = mon->vm;
+    bool incomplete = false;
+    int events = 0;
+    size_t sz = 0;
+
+    memset(buf + sz, 0, max_sz);
+    do {
+        ssize_t ret;
+
+        ret = read(mon->monitor_fd, buf + sz, max_sz - sz);
+        if (ret == 0 || (ret < 0 && errno == EINTR)) {
+            g_usleep(G_USEC_PER_SEC);
+            continue;
+        } else if (ret < 0) {
+            VIR_WARN("Failed to read monitor events!");
+            return -1;
+        }
+
+        sz += ret;
+        events = virCHMonitorValidateEventsJSON(mon->buffer,
+                        mon->buf_offset + sz, &incomplete);
+        VIR_DEBUG("Monitor event(%ld):\n%s", sz, mon->buffer);
+
+    } while (virDomainObjIsActive(vm) && (sz < max_sz) &&
+             (events == 0 || incomplete));
+
+    /*
+     * We process the events from the read buffer if
+     *    - There is atleast one event in the buffer
+     *    - No incomplete events in the buffer or
+     *    - The buffer is full and may have incomplete entries.
+     *
+     * If the buffer is full, virCHMonitorProcessEvents processes
+     * the completer events in the buffer and moves incomplete
+     * entries to the start of the buffer and next read from teh pipe
+     * starts from the offset.
+     */
+    mon->buf_fill_sz = sz + mon->buf_offset;
+    return (events > 0) ? virCHMonitorProcessEvents(mon, events) : events;
+}
+
 static void virCHMonitorEventLoop(void *data)
 {
     virCHMonitorPtr mon = (virCHMonitorPtr)data;
     virDomainObjPtr vm = mon->vm;
 
     VIR_DEBUG("Monitor event loop thread starting");
 
+    if (!(mon->buffer = g_malloc_n(sizeof(char), CH_MONITOR_BUFFER_SZ))) {
+        VIR_WARN("Failed to allocate monitor buffer!");
+        return;
+    }
+    mon->buf_offset = 0;
+    mon->buf_fill_sz = 0;
+
     virObjectRef(vm);
     while (g_atomic_int_get(&mon->event_loop_stop) == 0) {
-
-        g_usleep(G_USEC_PER_SEC);
-
-        virObjectLock(vm);
-        if (virDomainObjIsActive(vm) &&
-            virCHDomainObjBeginJob(vm, CH_JOB_MODIFY) == 0) {
-            virCHProcessSetupThreads(vm);
-            virCHDomainObjEndJob(vm);
-        }
-        virObjectUnlock(vm);
+        VIR_DEBUG("Reading events from monitor..");
+        if (virCHMonitorReadProcessEvents(mon) < 0)
+            VIR_WARN("Failed to process events from monitor!");
```

Comment:
Should we consider this fatal? It seems like not being able to talk with the monitor (or the monitor behaving poorly) could be considered inoperable/unrecoverable.

Reply:
Similar to your comment below, I think its better to start of fresh rather than considering it fatal. Ultimately, what I would like to see is a two way channel for the monitor-fd(like qemu). There should be a mechanism to let cloud-hypervisor know that consumers cannot reliably get events and cloud-hypervisor should start fresh sending data. Not sure how feasible this would be, but will discuss with Rob about this. Probably like an HTTP API call also would do.

But I feel that instead of bailing out, we should reset the event monitor buffer and then start fresh(discard all data until the next starting '{' is detected). What do you think?

Reply:
Yea, I like that approach. We just need to be sure we are keeping the buffer in a good state between calls in the eventloop while we are at it.

Needing to be sure we handle the cases of a full buffer being read but there not being a full event in the buffer so that we adjust everything so we can have the `{` at the start of the buffer for the next iteration of the eventloop.

Then we have to keep track of the case where we have a `{` at the start of the buffer and read the max size of our buffer but didn't see a full event. That case I'm not sure what the best way to handle is though, I'm open to suggestions.

Reply:
I have modified the code in such a way that no fatal situation exists after virCHMonitorReadProcessEvents returns